src/gui/windows/predictor/predictor_btb_dock.cpp

```diff
@@ -6,158 +6,158 @@ DockPredictorBTB::DockPredictorBTB(QWidget *parent) : Super(parent) {
-    connect(core, &machine::Core::step_started, this, &DockPredictorBTB::reset_colors);
+
+    clear();
 
     number_of_bits = init_number_of_bits(branch_predictor->get_number_of_btb_bits());
     const bool is_predictor_enabled { branch_predictor->get_enabled() };
 
     if (is_predictor_enabled) {
         btb->setRowCount(qPow(2, number_of_bits));
         btb->setDisabled(false);
+        clear_btb();
+
+        connect(
+            branch_predictor, &machine::BranchPredictor::btb_row_updated,
+            this, &DockPredictorBTB::update_btb_row);
+        connect(
+            branch_predictor, &machine::BranchPredictor::prediction_done,
+            this, &DockPredictorBTB::highligh_row_after_prediction);
+        connect(
+            branch_predictor, &machine::BranchPredictor::update_done,
+            this, &DockPredictorBTB::highligh_row_after_update);
+        connect(
+            core, &machine::Core::step_started,
+            this, &DockPredictorBTB::reset_colors);
+        
     } else {
         btb->setRowCount(0);
         btb->setDisabled(true);
     }
-    init_table();
-    btb->resizeRowsToContents();
-    set_table_color(Q_COLOR_DEFAULT);
-};
-
-void DockPredictorBTB::update_row(
-    uint16_t index,
-    machine::Address instruction_address,
-    machine::Address target_address) {
-    if (index >= btb->rowCount()) {
-        WARN("BTB dock update received invalid row index: %u", index);
+}
+
+void DockPredictorBTB::update_btb_row(
+    uint16_t row_index,
+    machine::BranchTargetBufferEntry btb_entry
+) {
+    if (row_index >= btb->rowCount()) {
+        WARN("BTB dock update received invalid row index: %u", row_index);
         return;
     }
 
-    set_row_color(index, Q_COLOR_UPDATE);
+    QTableWidgetItem *item;
 
-    QTableWidgetItem *item_index { btb->item(index, DOCK_BTB_COL_INDEX) };
-    QTableWidgetItem *item_instr_addr { btb->item(index, DOCK_BTB_COL_INSTR_ADDR) };
-    QTableWidgetItem *item_target_addr { btb->item(index, DOCK_BTB_COL_TARGET_ADDR) };
+    if (btb_entry.entry_valid) {
+        item = get_btb_cell_item(row_index, DOCK_BTB_COL_INSTR_ADDR);
+        item->setData(Qt::DisplayRole, machine::addr_to_hex_str(btb_entry.instruction_address));
 
-    if (item_index == nullptr) {
-        item_index = new QTableWidgetItem();
-        btb->setItem(index, DOCK_BTB_COL_INDEX, item_index);
-    }
+        item = get_btb_cell_item(row_index, DOCK_BTB_COL_TARGET_ADDR);
+        item->setData(Qt::DisplayRole, machine::addr_to_hex_str(btb_entry.target_address));
 
-    if (item_instr_addr == nullptr) {
-        item_instr_addr = new QTableWidgetItem();
-        btb->setItem(index, DOCK_BTB_COL_INSTR_ADDR, item_instr_addr);
-    }
+        item = get_btb_cell_item(row_index, DOCK_BTB_COL_TYPE);
+        item->setData(Qt::DisplayRole, machine::branch_type_to_string(btb_entry.branch_type).toString());
+    } else {
+        item = get_btb_cell_item(row_index, DOCK_BTB_COL_INSTR_ADDR);
```

Comment:
I would really prefer not to reuse to variable. It actually confused me for a while. Actually, you dont need a variable at all. You could just call `setData` on the result of the `get_...` method.

Reply:
Fixed